File: codes/golang/86.分隔链表.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {

	smlDummy, bigDummy := &ListNode{Val: 0}, &ListNode{Val: 0}
	sml, big := smlDummy, bigDummy

	for head != nil {
		if head.Val < x {
			sml.Next = head
			sml = sml.Next
		} else {
			big.Next = head
			big = big.Next
		}
		head = head.Next
	}

	sml.Next = bigDummy.Next
	big.Next = nil

	return smlDummy.Next
}

File: codes/rust/86.分隔链表.rs
impl Solution {
    pub fn partition(head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
        let mut left_dummy = Box::new(ListNode::new(0));
        let mut right_dummy = Box::new(ListNode::new(0));
        let mut left = &mut left_dummy;
        let mut right = &mut right_dummy;
        let mut head = head;
        while head.is_some() {
            let mut node = head.unwrap();
            head = node.next.take();
            if node.val < x {
                left.next = Some(node);
                left = left.next.as_mut().unwrap();
            } else {
                right.next = Some(node);
                right = right.next.as_mut().unwrap();
            }
        }
        // 将右边的链表追加的左边的链表中
        left.next = right_dummy.next.take();
        left_dummy.next
    }
}
